### DISCUSSION TITLE
`WasmEdge-plugin-wasi_nn-ggml-0.13.5-darwin_arm64.tar.gz` is built outside the CI?

### DISCUSSION THREAD

**Discussion by AkihiroSuda** [WasmEdge/WasmEdge#3126] [Q&A]
https://github.com/WasmEdge/WasmEdge/releases/download/0.13.5/WasmEdge-plugin-wasi_nn-ggml-0.13.5-darwin_arm64.tar.gz (sha256:a22b8a19fe43243dc7f4eb716f45e4ae95bf9375f2e0984df7720c007f9d14b5) doesn't seem built in the CI:
https://github.com/WasmEdge/WasmEdge/blob/c62e0bb3056bea6d26dab0e626de179cf0616243/.github/workflows/release.yml#L541-L658

From the result of the `strings`, it seems built in `/Users/hydai/workspace/WasmEdge/build_0.13.5_lts_b1656`.

Is there any reason for this?

**Comment by hydai** [comment]
All of the ggml plugins are built locally.

Its dependency, `llama.cpp`, updates quickly and is in active development. It may support more model types in just a week or less time. It's hard for us to release the WasmEdge with the same frequency.
To provide the latest `llama.cpp` support, we have a branch called `hydai/0.13.5_ggml_lts`, which will cherry-pick the `llama.cpp` related features. And we use this branch to build the ggml plugins for the various platforms.

Not only the Apple Silicon with Metal enabled built outside the CI, but also the following assets do:
1. x86_64 with CUDA support: We don't have a machine with a built-in GPU as a CI runner. Even if we build on CI, it is impossible to test and verify it.
2. aarch64 with CUDA support: the same reason as 1.

If you have any suggestions, feel free to let me know. We have already suffered from this problem for a pretty long time, and since we decided to support GPU-related features.

**Reply by AkihiroSuda** [reply]
Thanks for explanation.

> Even if we build on CI, it is impossible to test and verify it.

Would it be possible to build the binaries on GitHub Actions, upload them as GitHub Artifacts, and verify them locally before making a release?
That will help determining the origin and the compilation flags of the plugin binaries.

**Reply by juntao** [reply]
Yeah. I think to build in CI and then to test and release manually could be a way to go.

**Reply by hydai** [reply]
Sounds good. I would like to separate the ggml plugin from other extensions/wasi-nn plugins. The reason is that we may have more and more options that only apply to the ggml plugin. Said the CUDA-related dependencies, the Metal-related dependencies, and so on. Will have the new workflow soon. Thanks for your suggestion.
